staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -697,6 +704,12 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 		if int64(v.Len()) >= pred.Limit {
 			break
 		}
+		// There is still place in our bucket, but we have iterated too many elements not matching pred.
+		// To avoid returning timeout before we fill out the bucket, interrupt after maxKeysFetchedInSingleList elements
+		// and return continuation to resume scanning from the current position in the next request.
+		if allKeysCount >= maxKeysFetchedInSingleList {
+			break
```

Comment:
@liggitt - I've discussed this with Maciek and I'm not sure what to do with it.
Basically, with this change we can return a list with less than <limit> items together with continuation token. On one hand it's not a documented contract (so that's fine), otoh I'm wondering if someone doesn't rely on it (client-go doesn't seem to, but that's only a single example).

[BTW - if we agree it's fine, it might make sense to break it based on time (e.g. break after 50s or sth like that to no hit 1m default timeout), instead of breaking based on key count. But let's wait for Jordan`s (and others?) opinions whether we actually want to go this way before changing anything.]

@liggitt - WDYT?

Reply:
I think this changes the contract in a way that needs sufficient review and discussion that it should be a PR to the KEP after we have talked it out here, no matter what.

Forming some thoughts